community[minor]: Implement Doctran async execution

gopetracca opened the pull request:
**Description**

The DoctranTextTranslator has an async transform function that was not implemented because [the doctran library](https://github.com/psychic-api/doctran) uses a sync version of the `execute` method.

- I implemented the `DoctranTextTranslator.atransform_documents()` method using `asyncio.to_thread` to run the function in a separate thread.
- I updated the example in the Notebook with the new async version.
- The performance improvements can be appreciated when a big document is divided into multiple chunks.

Relates to:
- Issue #14645: https://github.com/langchain-ai/langchain/issues/14645
- Issue #14437: https://github.com/langchain-ai/langchain/issues/14437
- https://github.com/langchain-ai/langchain/pull/15264

gopetracca committed:
feat(DoctranTextTranslator): implement atransform_documents function

gopetracca committed:
feat: Add example of the DoctranTextTranslator async execution

gopetracca committed:
Merge branch 'master' into fix/14645/async-doctran

gopetracca committed:
Merge branch 'master' into fix/14645/async-doctran

gopetracca committed:
fix: reorder imports

gopetracca committed:
fix: linting issues

gopetracca committed:
Merge branch 'fix/14645/async-doctran' of github.com:gopetracca/langchain into fix/14645/async-doctran

gopetracca committed:
Merge branch 'master' into fix/14645/async-doctran

gopetracca committed:
fix: linting issues based on CI feedback

gopetracca committed:
fix(mypy): missing type annotations

gopetracca committed:
Merge branch 'master' into fix/14645/async-doctran

eyurtsev commented:
@gopetracca looks good overall.

Jupyter notebooks are async by default, so this snippet:

```python
# We need nest_asyncio to run async code in the Notebook
import asyncio

import nest_asyncio

nest_asyncio.apply()
```

is not correct. 

Would you mind removing it -- we generally want to avoid recommending `nest_asyncio` to users since it does monkey patching of the asyncio library and is not maintained anymore

eyurtsev committed:
Merge branch 'master' into fix/14645/async-doctran

gopetracca commented:
Hi @eyurtsev, thank you for your review. I modified the Jupyter notebook, and removed `nest_asyncio` and replaced with this:

```
import asyncio

loop = asyncio.get_event_loop()
task = loop.create_task(qa_translator.atransform_documents(documents))
result = await asyncio.gather(task)
print(result[0][0].page_content)
```

If you prefer, I can completely remove the Jupyter example and roll back to the previous version, so the PR will only include the async functionality. Let me know what you think.

gopetracca committed:
fix: remove nest_asyncio from Jupyter Notebook example.

gopetracca committed:
Merge branch 'fix/14645/async-doctran' of github.com:gopetracca/langchain into fix/14645/async-doctran

eyurtsev commented:
@gopetracca jupyter notebook already runs within an async event loop.

You can use an `await` keyword directly.

```python
result =  await qa_translator.atransform_documents(documents)
print(result[0][0].page_content)
```

gopetracca committed:
fix: simplify async code execution in Jupyter Notebook example.

gopetracca committed:
fix: from PR feedback.
- missing future annotations import
- doctran initialization

gopetracca commented on libs/community/langchain_community/document_transformers/doctran_text_translate.py:
```diff
@@ -36,10 +37,61 @@ def __init__(
 
+    async def _aparse_document(
+        self, doctran: Any, index: int, doc: Document
+    ) -> tuple[int, Any]:
```
Fixed in [b800962](https://github.com/langchain-ai/langchain/pull/22372/commits/b800962414e8b1251d34fb09893bde303b253d03)

gopetracca commented on libs/community/langchain_community/document_transformers/doctran_text_translate.py:
```diff
@@ -36,10 +37,61 @@ def __init__(
+            raise ImportError(
+                "Install doctran to use this parser. (pip install doctran)"
+            )
+        doctran = Doctran(
```
Fixed in [b800962](https://github.com/langchain-ai/langchain/pull/22372/commits/b800962414e8b1251d34fb09893bde303b253d03)

gopetracca commented:
@eyurtsev, thanks again for the review. I pushed the fixes.

gopetracca committed:
fix: lint Notebook

gopetracca commented:
Hi @eyurtsev, there was a lint issue with the modified notebook. I updated it.

gopetracca committed:
fix: python3.8 support

asyncio.to_thread is available from python 3.9+, so in order to support python3.8, the function is implemented as a private method of the class.

gopetracca commented:
Hi @eyurtsev, thanks for reviewing this.
[asyncio.to_thread](https://docs.python.org/3.9/library/asyncio-task.html#asyncio.to_thread) is available from Python 3.9+, so to support Python3.8, I implemented it as a private method within the class as recommended [here](https://stackoverflow.com/questions/65316863/is-asyncio-to-thread-method-different-to-threadpoolexecutor). 

If you prefer, I could wrap everything in a try/except block to catch the exception that will only be triggered on Python 3.8, and use the private method only in that case, but I think this way is cleaner.

eyurtsev commented on libs/community/langchain_community/document_transformers/doctran_text_translate.py:
```diff
@@ -65,3 +116,19 @@ def transform_documents(
+        Return a coroutine that can be awaited to get the eventual result of *func*.
+        """
+
+        loop = asyncio.get_event_loop()
```
Would using our built in functionality work here, so we don't end up with implementations per document transformer?

https://github.com/langchain-ai/langchain/blob/2aa2bdfa567baac18b4f41abf9e4b4712421b850/libs/core/langchain_core/runnables/config.py#L528-L528

```
return await run_in_execuctor(None, func, *args, **kwargs)
```
instead of _to_thread?

gopetracca replied:
Hi @eyurtsev, thanks! That makes a lot of sense. I couldn't find that implementation before. I refactored the code and it's cleaner now.

gopetracca committed:
refactor: use run_in_executor from langchain_core.runnables.config import run_in_executor

eyurtsev committed:
Merge branch 'master' into fix/14645/async-doctran
